Use live illink for trimming tests

sbomer opened the pull request:
Fixes #88624

sbomer committed:
Use live illink for trimming tests

sbomer committed:
Fix import path

radical commented:
I suggest running `runtime-wasm` also, which runs all the trimming, and AOT tests for wasm.

sbomer committed:
Fix NativeAot tests

sbomer committed:
Fix illink config for nativeaot tests

sbomer committed:
Prevent double file copy

ViktorHofer commented on eng/testing/linker/SupportFiles/Directory.Build.targets:
```diff
@@ -1,4 +1,5 @@
 <Project>
+  <Import Project="$(ToolsILLinkDir)build/Microsoft.NET.ILLink.Tasks.props" />
```
That file is expected to imported before the project body is evaluated. Please move this into a Directory.Build.props file so that we build the same as customers do via the PackageReference.

ViktorHofer commented on src/tools/illink/src/ILLink.RoslynAnalyzer/ILLink.RoslynAnalyzer.csproj:
```diff
@@ -9,6 +9,8 @@
     <AnalysisLevel>Latest</AnalysisLevel>
     <NoWarn Condition="'$(DotNetBuildFromSource)' == 'true'">$(NoWarn);CS8524</NoWarn>
     <AnalyzerLanguage>cs</AnalyzerLanguage>
+    <!-- This analyzer doesn't follow the convention for generator projects so we set the property manually. -->
+    <IsGeneratorProject>true</IsGeneratorProject>
```
Please undo this change. The `IsGeneratorProject` must be set in a Directory.Build.props file (if not inferred automatically based on the project file name): https://github.com/dotnet/runtime/blob/d4db2d96bb0f55211c30dc69ac587b204bb418f3/Directory.Build.props#L445

ViktorHofer commented on src/tools/illink/src/ILLink.Tasks/ILLink.Tasks.csproj:
```diff
@@ -20,6 +20,8 @@
          NU5129: Targets don't match the package id. This is intentional. -->
     <NoWarn>$(NoWarn);NU5128;NU5129</NoWarn>
     <IncludeBuildOutput>false</IncludeBuildOutput>
+    <!-- Place build output into a layout that matches the package layout. -->
+    <OutputPath>$(OutputPath)/tools</OutputPath>
```
The build and package output paths are intentionally not aligned. Why do you need the tools subdirectory for the output path?

ViktorHofer replied:
We should keep the output path and package layout separate, otherwise we will hit other issues over time. Please revert the `OutputPath` change.

sbomer committed:
PR feedback

- Move props import to top of project file
- Leave IsGeneratorProject in Directory.Build.props
  Ignore Directory.Build.props in Content

sbomer committed:
Merge remote-tracking branch 'origin/main' into liveILLink

sbomer committed:
Remove TFM condition

TFM isn't set during GenerateNuspec, but we need
Content to be defined so that the props/targets
get added to the package.

ViktorHofer commented on eng/testing/linker/project.csproj.template:
```diff
@@ -1,5 +1,12 @@
+    <ToolsILLinkDir>{ToolsILLinkDir}</ToolsILLinkDir>
+  </PropertyGroup>
+
+  <Import Project="$(ToolsILLinkDir)build/Microsoft.NET.ILLink.Tasks.props" />
```
Unfortunately this is still not the correct sequence point. The props file is expected to be imported before the project is evaluated, i.e. in the project's Directory.Build.props file. You can achieve that by setting the dependent `ToolsILLinkDir` property before an explicit `Sdk` import and then moving the new `Import` statement into the Directory.Build.props file.

sbomer committed:
Import from Directory.Build.props

sbomer committed:
Don't change outputpath

eerhardt commented on eng/testing/linker/project.csproj.template:
```diff
@@ -1,4 +1,11 @@
-<Project Sdk="Microsoft.NET.Sdk">
+<Project>
+
+  <PropertyGroup>
+    <!-- Needed for PublishTrimmed -->
+    <ToolsILLinkDir>{ToolsILLinkDir}</ToolsILLinkDir>
+  </PropertyGroup>
+
+  <Import Project="Sdk.props" Sdk="Microsoft.NET.Sdk" />
```
Do we really need to "break out" of using `Sdk="Microsoft.NET.Sdk"` here? Can we just move the `Directory.Build.props` change into this file (so all these "locate local tools" is contained in one place)?

sbomer committed:
Remove ToolsArtifactsPath

eerhardt commented on eng/pipelines/runtime.yml:
```diff
@@ -288,7 +288,7 @@ extends:
             timeoutInMinutes: 120
             nameSuffix: NativeAOT
-            buildArgs: -s clr.aot+host.native+libs -rc $(_BuildConfig) -lc Release -hc Release
+            buildArgs: -s clr.aot+host.native+libs+tools.illink -c $(_BuildConfig) -rc $(_BuildConfig) -lc Release -hc Release
```
What does -c do when we also specify -rc and -lc?

ViktorHofer replied:
The -c probably applies to the tools.illink subset.

ViktorHofer replied:
You could simplify this by removing the `-rc` switch.

eerhardt commented on eng/testing/linker/SupportFiles/Directory.Build.props:
```diff
@@ -1,4 +1,13 @@
+       the same layout as the NuGet package. -->
+  <Import Project="$(ToolsILLinkDir)build/Microsoft.NET.ILLink.Tasks.props" />
+  <PropertyGroup>
+    <ILLinkTasksAssembly>$(ToolsILLinkDir)net8.0/ILLink.Tasks.dll</ILLinkTasksAssembly>
```
Can we use CurrentTargetFramework here? So we don’t have to update net8.0 in net 9 time?

radical replied:
What is `CurrentTargetFramework`? I can't find that. I ask because I have an open PR with a similar change - https://github.com/dotnet/runtime/pull/89699 .

eerhardt replied:
https://github.com/dotnet/runtime/blob/30db9b78c8c21aea74a945332bda649e14b6b52d/src/libraries/Microsoft.Extensions.Caching.Abstractions/src/Microsoft.Extensions.Caching.Abstractions.csproj#L4

I meant NetCoreAppCurrent

vitek-karas replied:
I think this should use the `NetCoreAppToolCurrent` because that's what the target project uses: https://github.com/dotnet/runtime/blob/242242947c01615b243e8b4d9ee163030a4ceb47/src/tools/illink/src/ILLink.Tasks/ILLink.Tasks.csproj#L4

ViktorHofer replied:
Note that `$(TargetFramework)` shouldn't be used in a props file as that property isn't defined at that time for non multi-targeting projects. You could move this property into the Directory.Build.targets file.

ViktorHofer replied:
Alternatively, you could add the TFM as a property to the imported props file, to keep them in sync.

vitek-karas replied:
I wonder if we should make tools use the runtime configuration. For example when compiling AOT, the "tools" (ilc compiler) use runtime configuration. Additionally I ran into this myself and found it weird that specifying all of `-rc`, `-lc` and `-hc` wasn't enough, that there were still parts which didn't follow either of them.
(Obviously as a separate change)

ViktorHofer replied:
I asked about that when we brought the linker into runtime and the consensus was that the linker should not follow the runtime configuration, as its a separate tool but doesn't need to have yet another configuration property. The current configuration matrix is already quite complicated in dotnet/runtime.

vitek-karas replied:
Do we have other code which falls into the same bucket (as not belonging into any of the 3 configuration properties)?

ViktorHofer replied:
Here are the current configuration properties: https://github.com/dotnet/runtime/blob/11d9e87db295032a9927ee19bf3e2918539d8941/eng/Subsets.props#L548-L553

> Do we have other code which falls into the same bucket (as not belonging into any of the 3 configuration properties)?

The shared framework package projects: "packs". But those don't invoke the compiler.

ViktorHofer replied:
> I don't think this works because the imported file ships with the package (so needs to hard code the package TFM) - unless I misunderstood the suggestion.

You already hardcode the TFM in the shipping props file: https://github.com/dotnet/runtime/blob/e43cba31e997e960205844b3f0541011ef732c29/src/tools/illink/src/ILLink.Tasks/build/Microsoft.NET.ILLink.Tasks.props#L21-L22

I was saying that you could introduce another property in that file that just includes the `net8.0\ILLink.Tasks.dll` part that you could then use here to keep the TFM in sync.

sbomer committed:
Use NetCoreAppToolCurrent
